### DISCUSSION TITLE
Memory efficient multi sink

### DISCUSSION THREAD

**Discussion by dloebl** [libvips/libvips#3184] [Ideas]
We need the ability to output multiple image formats from just one input image (in sequential access). For instance:
`PNG -> resize -> output WebP -> output AVIF`

Right now, our solution is to call [vips_image_wio_input()](https://www.libvips.org/API/current/VipsImage.html#vips-image-wio-input) on the image before the calls to e.g. `vips_webpsave()`, `vips_heifsave()` etc.

**Simplified example:**
```
  VipsImage* a, *b;
  a = vips_image_new_from_file("test.png", "access", VIPS_ACCESS_SEQUENTIAL, NULL);
  vips_resize(a, &b, 0.5, NULL);
  vips_image_wio_input(b);
  vips_pngsave(b, "out.png", NULL);
  vips_jpegsave(b, "out.jpg", NULL);
```

Is it possible to write to multiple output targets in a fast and efficient way (in sequential access) without requiring to buffer the complete image into memory or would this kind of feature even be applicable to libvips?

Related discussion: https://github.com/libvips/libvips/discussions/2923

**Comment by jcupitt** [comment]
Hi @dloebl,

This would be a nice thing to have, but it doesn't fit with demand-driven evaluation, unfortunately. As you say, you need to render the whole image to memory and then you can run several writes from that.

I'd use `copy_memory()` rather than `wio_input()` -- `wio_input` modifies its argument, so it's not safe to use on a descriptor that's shared between threads.

You can run the save operations in parallel (I expect you know).
